CI enhancements - Update Dockerfile to working version, update build+fmt+imports

thomasmodeneis opened the pull request:
## Changes
- fixing Dockerfile.dev to build properly
- update Makefile tasks and fix typos
- add ci go file for ci builds to it works also on non bash
- add gmfmt and goimports task to Makefile and bash script
- 

## Tests:
```
- make gossamer 
will build local gossamer to `build/bin/gossamer`

- make docker
will build and start gossamer docker image


```

### Issues:

ansermino commented on interfaces.go:
```diff
@@ -0,0 +1 @@
+package gossamer
```
Please could you explain why this needed.

thomasmodeneis replied:
to avoid building errors like: 
`can't load package: package .: no Go files in /gocode/src/github.com/ChainSafe/gossamer`

ansermino replied:
I don't see how this changes anything. Please explain or remove

thomasmodeneis replied:
If I remove, the project wont compile.

ansermino replied:
`go build` works fine when called from within `cmd/gossamer`. Running `go build` from the root is incorrect as there is no root package. `go build github.com/ChainSafe/gossamer/cmd/gossamer` should work from anywhere.

thomasmodeneis replied:
yes, correct. I managed to remove it and build works fine now. Cheers.

ansermino commented on build/ci.go:
```diff
@@ -0,0 +1,61 @@
+	}
+}
+
+func install() {
```
Could you explain this a bit? I don't understand why you run `go install` on every package.

thomasmodeneis replied:
that is to avoid issues with weird compilation issues like missing packages / deps.

ansermino replied:
I'm not sure I understand. Is there a particular scenario this would fix?

thomasmodeneis replied:
Sure, it will avoid several issues with c modules or cgo issues. The Dockerfile was broken, I first tried to update to last go13 and latest Apline version. I spend a whole afternoon working to fix issues, and mostly related to eg: "could not compile crypto" or "package crypto not found" or missing c libs related to the wasm, or gcc missing, or g++ missing, or other deps that could not be found or compiled. So I added the ci.go into it to avoid the issues. Besides, it will alow people on other distros (eg: windows) to compile it properly without having to dive into go install issues.

ansermino replied:
I'm still not sure how this would fix that. Running `go install` on every package should not be required to build this. Can you share any further details about the errors you ran into? Perhaps you can open an issue?

thomasmodeneis replied:
well when you do go install ./... you are doing the same thing but you are mode dependent on the env configurations. When you use go to build go, you can then be sure about the packages are being build the same independently of the OS. In any case, being explicit on the packages and have output on the console is always useful for CI multi cross package builds on go. 

I can share the issues with it, perhaps to try to fix the Alpine build in the future.

thomasmodeneis replied:
I've raised https://github.com/ChainSafe/gossamer/issues/424
I would still like to keep the ci.go build because I am not sure if Windows compilation actually works. Please let me know if you want me to ditch it so I can do it and go ahead with the issue.
Cheers.

ansermino replied:
Check out my comment on #424

thomasmodeneis replied:
Just replied there, I really would like to have this in asap because its kind of blocking my local build as the `make gossamer` and `make start` are currently not really nice to use :)
Let me know what is your decision, I can remove the go building go if this is what you want.
Cheers.

ansermino replied:
i think I'm starting to see the appeal of this. Certainly as we move to more complex testing configurations this will be useful.

Thanks for your patience, apologies for how this has been dragged out.

thomasmodeneis replied:
no worries, its my pleasure to help :)
--
I did the requested changes. Would you like to create a new issue for adding tests into ci.go in a near future ?

ansermino replied:
Yeah that would be great!

thomasmodeneis commented:
> Awesome work, thank you!
> 
> The dockerfile is missing one small detail -- in order to start the node you need to init the genesis state. You need `genesis.json`, then you can run `gossamer init` which will use the default path `./genesis.json`.

I just added the docker entrypoint script and updated the Dockerfile and now it will init genesis if not yet initialized then start the node.
I also updated the Makefile docker tasks and separated the compile tasks from run task.

Cheers.

dutterbutter commented:
Hey, @thomasmodeneis great job on this PR and thank you very much for your contribution!

Everything looks good to me! If you don't mind please add docker build job in `.travis.yml` and we can get this PR merged. 

🔥🔥🔥

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ChainSafe/gossamer?pullRequest=409) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ChainSafe/gossamer?pullRequest=409) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ChainSafe/gossamer?pullRequest=409) it.</sub>

ansermino commented on build/ci.go:
```diff
@@ -0,0 +1,61 @@
+	cmd.Args = append(cmd.Args, "-v")
+	cmd.Args = append(cmd.Args, packages...)
+
+	fmt.Println("Build Gossamer", strings.Join(cmd.Args, " "))
```
This is a bit ugly, I think `" \\\n"` would be better as the separator.

thomasmodeneis replied:
cool, sure

ansermino commented on build/ci.go:
```diff
@@ -0,0 +1,61 @@
+// +build none
```
can we move this to `scripts/`? not a fan of having a non-disposible `build` directory

thomasmodeneis replied:
sure
